### ISSUE TITLE
dflib-jupyter and echarts

### ISSUE THREAD

**Issue by maxandersen** [dflib/dflib#410]
trying to get echarts working with dflib: https://github.com/jupyter-java/jupyter-java-examples/blob/main/notebooks/java-dflib-echarts.ipynb

```
import org.dflib.echarts.*;

DataFrame df = DataFrame.foldByRow("name", "salary").of(
                "J. Cosin", 120000,
                "J. Walewski", 80000,
                "J. O'Hara", 95000)
        .sort($col("salary").desc());

var chart = ECharts
        .chart()
        .xAxis("name")
        .series(SeriesOpts.ofBar(), "salary")
        .plot(df);

        chart;
```

See that it does generates a html/text output cell with:

```
<script type='text/javascript' src='https://cdn.jsdelivr.net/npm/echarts@5.5.1/dist/echarts.min.js'></script>
<div id='dfl_ech_98' style='width: 600px;height:400px;'></div>
<script type='text/javascript'>
    var chart_dfl_ech_98 = echarts.init(
        document.getElementById('dfl_ech_98'),
        null,
   );
    var option_dfl_ech_98 = {
        dataset: {
            source: [
                ['L0','J. Cosin','J. O\'Hara','J. Walewski'],
                ['salary',120000,95000,80000]
            ]
        },
        xAxis: [
        {
            type: 'category'
        },
        ],
        yAxis: [
        {
            type: 'value'
        },
        ],
        series: [
            {
                name: 'salary',
                encode: {
                    x: 0,
                    y: 1,
                },
                seriesLayoutBy: 'row',
                type: 'bar'
            },
        ]
    };
    option_dfl_ech_98 && chart_dfl_ech_98.setOption(option_dfl_ech_98);</script>
```

but i cant get that to render in vscode jupyter notebook nor in github notebook preview.

**Comment by andrus** [comment]
I am able to reproduce it partially. It is that "race condition" that I mentioned elsewhere. The first time the output is empty box. The second time I rerun the cell, it works ok (see below). That's in `jupyter lab` in Firefox. Let me try vscode jupyter 


<img width="789" alt="image" src="https://github.com/user-attachments/assets/64d32e81-cd64-4dbf-a293-ee272d6ab9a5" />

**Comment by andrus** [comment]
Yes, def some kind of race condition. Every now and then it renders in VSCode, but that's entirely unpredictable. We'll investigate 
<img width="811" alt="image" src="https://github.com/user-attachments/assets/232f27a7-4bfc-4fcf-8806-b0e70164974b" />

**Comment by andrus** [comment]
From a simple fix this turned into something much bigger. So...:

1. `jupyter lab` in the browser.  Charts work. Very rarely may require to re-execute a cell. We have a fix to make it more robust, but can't even reproduce the original issue as it is not that common.
2. GitHub doesn't allow even the simplest local JS from a notebook output to be executed. 
3. VSCode can exec JS and can load external JS files. But the rendering doesn't work most of the time (it does sometimes).

The reason Kotlin Kandy works in cases 2 & 3 is that the rendering seems to be done server-side and there are embeded SVGs in the `.ipynb`. Our rendering is 100% client-side, so we need JS to work. So GitHub will continue to show an empty cell (actually a properly sized div, just without a chart  🙂 ) until we can switch to server-side rendering somehow. I am not giving up on VSCode just yet, but need to figure out how to debug JS execution within the notebook (e.g. how do I check the console logs)

**Comment by andrus** [comment]
Ok, getting somewhere with VSCode. There's a ["Jupyter Notebook Renderers"](Jupyter Notebook Renderers) extension from Microsoft. If I disable it and restart VSCode, the charts start rendering. I suspect it loads some conflicting JS libraries will need further investigation, but at least we know where to look

<img width="717" alt="image" src="https://github.com/user-attachments/assets/c0f4193d-f17d-4b75-ba9e-5037bc5453eb" />

**Comment by maxandersen** [comment]
can confirm that disabling that notebook renderer extension makes it work for me.

that is unfortunate as it provides lots of other nice renderers.

**Comment by maxandersen** [comment]
whats weird though is that notebook renderer isn't supposed to affect text/html...

**Comment by maxandersen** [comment]
using developer tools I took a snapshot to spot the diff with and without the rendering.

With jupyter notebook rendering there are *two* iframes - one nested in another.

```
<iframe title="Notebook webview content" id="active-frame" frameborder="0" sandbox="allow-same-origin allow-pointer-lock allow-scripts allow-downloads allow-forms" allow="cross-origin-isolated; autoplay; clipboard-read; clipboard-write;" src="./fake.html?id=e072ef26-2ba7-4d84-b3ad-fb750abe502c" style="display: block; margin: 0px; overflow: hidden; position: absolute; width: 100%; height: 100%; visibility: visible;"><iframe title="Notebook webview content" id="active-frame" frameborder="0" sandbox="allow-same-origin allow-pointer-lock allow-scripts allow-downloads allow-forms" allow="cross-origin-isolated; autoplay; clipboard-read; clipboard-write;" src="./fake.html?id=e072ef26-2ba7-4d84-b3ad-fb750abe502c" style="display: block; margin: 0px; overflow: hidden; position: absolute; width: 100%; height: 100%; visibility: visible;"></iframe></iframe></body>
```

without its just one iframe and all is good.

```
<iframe title="Notebook webview content" id="active-frame" frameborder="0" sandbox="allow-same-origin allow-pointer-lock allow-scripts allow-downloads allow-forms" allow="cross-origin-isolated; autoplay; clipboard-read; clipboard-write;" src="./fake.html?id=e3ba66ab-defd-4f6f-b7f8-2685594c9154" style="display: block; margin: 0px; overflow: hidden; position: absolute; width: 100%; height: 100%; visibility: visible;"></iframe></body>
```

**Comment by andrus** [comment]
Aha. I am not much of a VSCode user, so I was unsure how to even debug the HTML/JS involved. Thanks for the hint about developer tools. We'll see if we can do anything about it from our end.

**Comment by Vladiimir** [comment]
I've made investigation of echarts rendering issue in vscode. I would like to share my conclusions.
When we download js with html tag `<script src=".../echarts.js">` echarts object is stored in the window global object.
But when Jupyter Notebook Renderers plugin is enabled it affects window object and delete window.echarts property. It causes the error.
If use requireJS for loading echarts.js it loads script to local context and and after that i can set it to window.echarts property and this manipulation allows to walk around conflict with renderers plugin 
` // Config for RequireJS
      require.config({
           paths: {
                'echarts': 'https://cdn.jsdelivr.net/npm/echarts@5.5.1/dist/echarts.min'
            }
      });
    
        // Using require.js for loading echarts
        require(['echarts'], function(echarts) {
                window.echarts = echarts;
                ........init echarts and setOption............
       });
`
For now I'm trying to find some solution that doesn't need additional js library as requireJS is.

**Comment by Vladiimir** [comment]
this script works with Notebook Renderers Plugin and without it.

`

"text/html": [
       "<div id='dfl_ech_6877' style='width: 600px;height:400px;'></div>\n",
        "<script type='text/javascript'>\n",
        "function renderChart() {",
        "    var chart_dfl_ech_6877 = echarts.init(",
        "        document.getElementById('dfl_ech_6877'),\n",
        "        null\n",
        "    );\n",
        "    var option_dfl_ech_6877 = {\n",
        "        dataset: {\n",
        "            source: [\n",
        "                ['L0','J. Cosin','J. O\\'Hara','J. Walewski'],\n",
        "                ['salary',120000,95000,80000]\n",
        "            ]\n",
        "        },\n",
        "        xAxis: [\n",
        "            {\n",
        "                type: 'category'\n",
        "            }\n",
        "        ],\n",
        "        yAxis: [\n",
        "            {\n",
        "                type: 'value'\n",
        "            }\n",
        "        ],\n",
        "        series: [\n",
        "            {\n",
        "                name: 'salary',\n",
        "                encode: {\n",
        "                    x: 0,\n",
        "                    y: 1\n",
        "                },\n",
        "                seriesLayoutBy: 'row',\n",
        "                type: 'bar'\n",
        "            }\n",
        "        ]\n",
        "    };\n",
        "    option_dfl_ech_6877 && chart_dfl_ech_6877.setOption(option_dfl_ech_6877);\n",
        "}",
        "if (window.require) {",
        "    console.log('Property require is in window.');",
        "    require.config({",
        "        paths: {",
        "            'echarts': 'https://cdn.jsdelivr.net/npm/echarts@5.5.1/dist/echarts.min'\n",
        "        }\n",
        "    });\n",
        "    require(['echarts'], function(echarts) {",
        "        window.echarts = echarts;",
        "        renderChart();",
        "    });\n",
        "} else {",
        "    var script = document.createElement('script');",
        "    script.type = 'text/javascript';",
        "    script.src = 'https://cdn.jsdelivr.net/npm/echarts@5.5.1/dist/echarts.min.js';",
        "    script.addEventListener('load', function() {",
        "        console.log('ECharts is loaded and available.');",
        "        renderChart();",
        "    });",
        "    document.head.appendChild(script);",
        "}",
        "</script>"

      ],

`

**Comment by andrus** [comment]
For the record, the new bootstrap script per #574 works fine in VSCode (including theme loading), but the issue with the "Jupyter Notebook Renderers extension remains unresolved.

**Comment by andrus** [comment]
We were finally able to get to the bottom of it! Thanks @Vladiimir. Charts are working for me now in VSCode with "Jupyter Notebook Renderers" extension and all other environments. @maxandersen , let us know if you still see any unexpected behavior in VSCode.
